# 1. Import the necessary libraries and read the provided CSVs as a data frame 

In [29]:
import pandas as pd
import os
import glob
from surprise import SVD
from surprise import KNNWithMeans
from surprise import Dataset
from surprise import accuracy
from surprise import Reader
from surprise.model_selection import cross_validate
from surprise.model_selection import train_test_split
from collections import defaultdict
import seaborn as sns

1.A. Merge all the provided CSVs into one dataFrame.

In [2]:
path = os.getcwd()
path

'/Users/gobinathvelusamy'

In [3]:
csv_files = glob.glob(os.path.join('data/', "*.csv"))

In [4]:
csv_files

['data/phone_user_review_file_1.csv',
 'data/phone_user_review_file_2.csv',
 'data/phone_user_review_file_3.csv',
 'data/phone_user_review_file_6.csv',
 'data/phone_user_review_file_4.csv',
 'data/phone_user_review_file_5.csv']

In [5]:
for f in csv_files:
    df = pd.read_csv(f,encoding='latin-1')

B2.. Explore, understand the Data and share at least 2 observations. 

In [6]:
df.shape

(350216, 11)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350216 entries, 0 to 350215
Data columns (total 11 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   phone_url  350216 non-null  object 
 1   date       350216 non-null  object 
 2   lang       350216 non-null  object 
 3   country    350216 non-null  object 
 4   source     350216 non-null  object 
 5   domain     350216 non-null  object 
 6   score      321983 non-null  float64
 7   score_max  321983 non-null  float64
 8   extract    341836 non-null  object 
 9   author     321351 non-null  object 
 10  product    350216 non-null  object 
dtypes: float64(2), object(9)
memory usage: 29.4+ MB


In [8]:
df.describe()

score  score_max
count  321983.000000   321983.0
mean        7.773498       10.0
std         2.644970        0.0
min         0.200000       10.0
25%         6.000000       10.0
50%         8.000000       10.0
75%        10.000000       10.0
max        10.000000       10.0

In [9]:
df.head()

phone_url       date lang country      source  \
0  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
1  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
2  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
3  /cellphones/karbonn-k1616/  4/25/2014   en      in     Naaptol   
4  /cellphones/karbonn-k1616/  4/23/2013   en      in     Naaptol   

          domain  score  score_max  \
0  91mobiles.com    2.0       10.0   
1  91mobiles.com    6.0       10.0   
2  91mobiles.com    4.0       10.0   
3    naaptol.com   10.0       10.0   
4    naaptol.com   10.0       10.0   

                                             extract           author  \
0  I bought 1 month before. currently speaker is ...        venkatesh   
1  I just bought one week back, I have Airtel con...           Venkat   
2  one problem in this handset opera is not worki...           krrish   
3  here Karbonn comes up with an another excellen...  BRIJESH CHAUHAN   
4  What a phone, all so on Naaptol my god 23% off...    Suraj CHAUHAN   

                 product  
0          Karbonn K1616  
1          Karbonn K1616  
2          Karbonn K1616  
3  Karbonn K1616 - Black  
4  Karbonn K1616 - Black

Observation : 
1.Data set has 350216 rows and 11 Columns 
2.Score & Score_max columns has integer data type remaning Columns are categorical 
3.Score_max has uunique value 10 . So std deviationas Zero
4.Mean values of Score will be 7.773498

1.C. Round off scores to the nearest integers. 

In [11]:
df[['score','score_max']] = df[['score','score_max']].round().astype('Int64')

In [12]:
pd.set_option("display.max_rows", None)

In [13]:
df.head()

phone_url       date lang country      source  \
0  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
1  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
2  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
3  /cellphones/karbonn-k1616/  4/25/2014   en      in     Naaptol   
4  /cellphones/karbonn-k1616/  4/23/2013   en      in     Naaptol   

          domain  score  score_max  \
0  91mobiles.com      2         10   
1  91mobiles.com      6         10   
2  91mobiles.com      4         10   
3    naaptol.com     10         10   
4    naaptol.com     10         10   

                                             extract           author  \
0  I bought 1 month before. currently speaker is ...        venkatesh   
1  I just bought one week back, I have Airtel con...           Venkat   
2  one problem in this handset opera is not worki...           krrish   
3  here Karbonn comes up with an another excellen...  BRIJESH CHAUHAN   
4  What a phone, all so on Naaptol my god 23% off...    Suraj CHAUHAN   

                 product  
0          Karbonn K1616  
1          Karbonn K1616  
2          Karbonn K1616  
3  Karbonn K1616 - Black  
4  Karbonn K1616 - Black

1.D. Check for missing values. Impute the missing values, if any. 

In [14]:
df.isnull().sum() / df.shape[0] * 100.00

phone_url    0.000000
date         0.000000
lang         0.000000
country      0.000000
source       0.000000
domain       0.000000
score        8.061596
score_max    8.061596
extract      2.392809
author       8.242056
product      0.000000
dtype: float64

In [15]:
# filling the null values in column 'score' and 'score_max' as these are numerical Values
df = df.fillna(df.median())

/var/folders/7p/5k1s0qxj2mn1kmk2d_npxq8r0000gn/T/ipykernel_1117/885693151.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df = df.fillna(df.median())


In [16]:
df.isnull().sum() / df.shape[0] * 100.00

phone_url    0.000000
date         0.000000
lang         0.000000
country      0.000000
source       0.000000
domain       0.000000
score        0.000000
score_max    0.000000
extract      2.392809
author       8.242056
product      0.000000
dtype: float64

In [17]:
# dropping the null values in columns 'extract' ,'author' and 'product'
df = df.dropna()

In [18]:
df.isnull().sum() / df.shape[0] * 100.00

phone_url    0.0
date         0.0
lang         0.0
country      0.0
source       0.0
domain       0.0
score        0.0
score_max    0.0
extract      0.0
author       0.0
product      0.0
dtype: float64

1.E.Check for duplicate values and remove them, i

In [19]:
duplicate = df[df.duplicated()]
duplicate

phone_url        date lang  \
866                            /cellphones/alcatel-8000d/   7/26/2014   ru   
885                            /cellphones/alcatel-8000d/   2/11/2014   ru   
1411                   /cellphones/samsung-galaxy-pocket/  12/30/2012   it   
2156                                 /cellphones/lg-a290/   6/26/2013   pt   
2382               /cellphones/prestigio-multiphone-4000/   3/13/2014   ru   
2391               /cellphones/prestigio-multiphone-4000/   7/18/2013   ru   
2396               /cellphones/prestigio-multiphone-4000/   6/13/2013   ru   
3384                          /cellphones/nokia-asha-203/   4/15/2013   ru   
3451                          /cellphones/nokia-asha-203/  12/19/2012   it   
3455                          /cellphones/nokia-asha-203/  11/14/2012   it   
3477                          /cellphones/nokia-asha-203/    8/7/2012   ru   
3802                      /cellphones/lg-optimus-l3-e405/   4/25/2014   ru   
3862                      /cellphones/lg-optimus-l3-e405/   1/12/2014   pt   
3935                      /cellphones/lg-optimus-l3-e405/   11/8/2013   pt   
4491                      /cellphones/lg-optimus-l3-e405/   1/22/2013   pt   
4517                      /cellphones/lg-optimus-l3-e405/    1/7/2013   pt   
4546                      /cellphones/lg-optimus-l3-e405/  12/30/2012   ru   
4630                      /cellphones/lg-optimus-l3-e405/  11/28/2012   ru   
4738                      /cellphones/lg-optimus-l3-e405/  10/22/2012   ru   
4809                      /cellphones/lg-optimus-l3-e405/   9/16/2012   ru   
5052                      /cellphones/lg-optimus-l3-e405/   7/15/2012   ru   
5056                      /cellphones/lg-optimus-l3-e405/   7/14/2012   ru   
5187                      /cellphones/lg-optimus-l3-e405/   3/11/2012   ru   
5192                      /cellphones/lg-optimus-l3-e405/   2/11/2012   ru   
5657                           /cellphones/samsung-c3520/   8/10/2014   ru   
6085                           /cellphones/samsung-c3520/  11/11/2012   it   
6618                          /cellphones/alcatel-ot-536/  11/16/2013   ru   
6635                          /cellphones/alcatel-ot-536/   3/10/2013   ru   
6666                          /cellphones/alcatel-ot-536/    7/6/2012   ru   
7106                         /cellphones/nokia-lumia-510/   12/7/2014   it   
7121                         /cellphones/nokia-lumia-510/    7/9/2014   ru   
7330                           /cellphones/samsung-e2600/    5/2/2012   ru   
7341                           /cellphones/samsung-e2600/   3/11/2012   ru   
7377          /cellphones/samsung-star-deluxe-duos-s5292/   7/30/2014   ru   
7741                           /cellphones/lg-optimus-l7/  10/26/2014   pt   
8512                           /cellphones/lg-optimus-l7/   12/8/2012   ru   
8530                           /cellphones/lg-optimus-l7/  11/29/2012   ru   
8531                           /cellphones/lg-optimus-l7/  11/29/2012   ru   
8590                           /cellphones/lg-optimus-l7/  11/12/2012   it   
8656                           /cellphones/lg-optimus-l7/  10/11/2012   ru   
8844                           /cellphones/lg-optimus-l7/   8/24/2012   ru   
8846                           /cellphones/lg-optimus-l7/   8/23/2012   ru   
8859                           /cellphones/lg-optimus-l7/   8/21/2012   ru   
8882                           /cellphones/lg-optimus-l7/   8/14/2012   ru   
8904                           /cellphones/lg-optimus-l7/   7/31/2012   ru   
9233                           /cellphones/lg-optimus-l7/   1/11/2012   ru   
9297                       /cellphones/sony-xperia-neo-l/   2/18/2014   ru   
9299                       /cellphones/sony-xperia-neo-l/   1/22/2014   ru   
9304                       /cellphones/sony-xperia-neo-l/  10/16/2013   ru   
9911                         /cellphones/caterpillar-b25/  12/14/2013   ru   
9956                         /cellphones/caterpillar-b25/   7/31/2013 

We have 2160 Duplicated Records 

In [20]:
df=df.drop_duplicates(keep=False)

In [21]:
duplicate = df[df.duplicated()]
duplicate

Empty DataFrame
Columns: [phone_url, date, lang, country, source, domain, score, score_max, extract, author, product]
Index: []

1.F. Keep only 1 Million data samples. Use random state=612.

In [22]:
df1 = df.sample(n=1000000, random_state=612,replace=True)

In [23]:
df1.shape

(1000000, 11)

In [24]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000000 entries, 186305 to 346274
Data columns (total 11 columns):
 #   Column     Non-Null Count    Dtype 
---  ------     --------------    ----- 
 0   phone_url  1000000 non-null  object
 1   date       1000000 non-null  object
 2   lang       1000000 non-null  object
 3   country    1000000 non-null  object
 4   source     1000000 non-null  object
 5   domain     1000000 non-null  object
 6   score      1000000 non-null  Int64 
 7   score_max  1000000 non-null  Int64 
 8   extract    1000000 non-null  object
 9   author     1000000 non-null  object
 10  product    1000000 non-null  object
dtypes: Int64(2), object(9)
memory usage: 93.5+ MB


 G. Drop irrelevant features. Keep features like Author, Product, and Score.

In [25]:
df2=df1.drop(['phone_url','date','lang','country','source','domain','score_max','extract'],axis=1)

In [26]:
df2.head(10)

score               author  \
186305      8             chuqueno   
297238      8                dunya   
234982     10         Amazon Kunde   
207714      2       Evren Yurtesen   
274902      9           kimmy4vols   
314871      2        Luis ogando C   
303659      6  ANDERSONETALITA2010   
14098       8                Wand    
309754     10          rammstein95   
314737     10            SERGIOV89   

                                                  product  
186305                                         NOKIA 7230  
297238                 Blackberry Storm 9530 Cep Telefonu  
234982  Microsoft Nokia 5800 XpressMusic Comes with Mu...  
207714  Motorola Milestone 2 Unlocked Cell Phone with ...  
274902                                            VX-8360  
314871  Samsung Galaxy S II i777 16GB GSM Android Smar...  
303659                               BlackBerry Bold 9000  
14098                                 LG P690 Optimus Net  
309754                                 Nokia 3120 Classic  
314737                                           LG KP265

# 2 Part 

2.A. Identify the most rated products. 

In [34]:
df2.groupby('product')['score'].mean().sort_values(ascending=False).head()  

product
Nokia E63 Bleu outremer Symbian OS 9.2 / Series 60 3.1 Edition                                                                     10.0
BlackBerry Curve 9300 Zwart                                                                                                        10.0
Sony Ericsson LT15IEUBLU XPERIA Arc Android Smartphone with Touchscreen, 8MP Camera, Bluetooth, Wi-Fi, GPS - No Warranty - Blue    10.0
HTC HD2 Zwart                                                                                                                      10.0
The Sony Ericsson Xperia X10                                                                                                       10.0
Name: score, dtype: Float64

2.B Identify the users with most number of reviews.

In [63]:
(df2['author'].value_counts()).head()

Amazon Customer    12977
Anonymous           3121
e-bit               2289
Anonyme             1241
Гость          1212
Name: author, dtype: int64

C. Select the data with products having more than 50 ratings and users who have given more than 50 ratings.

In [65]:
df2['product'].value_counts().head()

Nokia 5800 XpressMusic    5181
Nokia 5800                5002
Nokia N8                  4602
LG GX200                  4565
Nokia E71                 3918
Name: product, dtype: int64

In [66]:
#extracting authors who gave greater than 50 ratings
df_a = pd.DataFrame(columns=['author', 'a_count'])

In [67]:
df_a

Empty DataFrame
Columns: [author, a_count]
Index: []

In [68]:
df2.shape

(1000000, 3)

In [69]:
df_a['author']=df2['author'].value_counts().index.tolist() 
df_a['a_count'] = list(df2['author'].value_counts() > 50)

In [70]:
# get names of indexes for which count column value is False
index_names = df_a[ df_a['a_count'] == False ].index 
# drop these row indexes from dataFrame 
df_a.drop(index_names, inplace = True) 
df_a

author  a_count
0                                      Amazon Customer     True
1                                            Anonymous     True
2                                                e-bit     True
3                                              Anonyme     True
4                                           Гость     True
5                                              Misafir     True
6                                         Amazon Kunde     True
7                                           zehrabalta     True
8                                            kuzeykizi     True
9                                                 Alex     True
10                                           ceren2006     True
11                                              retkit     True
12                                  Александр     True
13                                                   #     True
14                                        Сергей     True
15                                                kyef     True
16                                             ozerbjk     True
17                                               Chris     True
18                                      Дмитрий     True
19                                                ozer     True
20                                          miatamania     True
21                                           emanetcan     True
22                                              huliye     True
23                                              Daniel     True
24                                        Андрей     True
25                                             bumudur     True
26                                      Алексей     True
27                                        Amazon-Kunde     True
28                                             Michael     True
29                                               David     True
30                                     Kindle Customer     True
31                                               black     True
32                                           dilemma21     True
33                                              okuyan     True
34                                            zeynep34     True
35                                   anonieme bezoeker     True
36                                          ozlemce108     True
37                                      Cliente Amazon     True
38                                              anneme     True
39                                            gumusoje     True
40                                              Thomas     True
41                                               Peter     True
42                                             yusuf60     True
43                                                John     True
44                                             osmntyr     True
45                                              Andrea     True
46                                              Qantas     True
47                                              iboibo     True
48                                                Mike     True
49                                               dunya     True
50                                           Christian     True
51                                              sezenm     True
52                                      Евгений     True
53                                            Anônimo     True
54                                                Paul     True
55                                          cinderella     True
56                                              anonym     True
57                                               Marco     True
58                                              Stefan     True
59                                              gulayy     True
60                                          Игорь     True
61                                              Martin     True
62              

In [71]:
df_p = pd.DataFrame(columns=['product', 'p_count'])
df_p['product']=df2['product'].value_counts().index.tolist() 
df_p['p_count'] = list(df2['product'].value_counts() > 50)

In [72]:
# get names of indexes for which count column value is False
index_names = df_p[ df_p['p_count'] == False ].index 
# drop these row indexes from dataFrame 
df_p.drop(index_names, inplace = True)


In [ ]:
df_p

In [73]:
df_p1 = df2[df2['product'].isin(df_p['product'])] 

In [74]:
df_a1 = df_p1[df_p1['author'].isin(df_a['author'])]

In [ ]:
df_a1

In [75]:
df_a1.shape

(87273, 3)

# 3. Build a popularity based model and recommend top 5 mobile phones

In [76]:
#calculating the mean score for a product by grouping it.
ratings_mean_count = pd.DataFrame(df2.groupby('product')['score'].mean()) 

In [77]:
# calculating the number of ratings a product got
ratings_mean_count['rating_counts'] = pd.DataFrame(df2.groupby('product')['score'].count())  

In [78]:
# 3. Recommending the 5 mobile phones based in highest mean score and highest number of ratings the product got. 
ratings_mean_count.sort_values(by=['score','rating_counts'], ascending=[False,False]).head()

score  rating_counts
product                                                                 
Nokia : Sí : Sí 6630                           10.0             55
Doro Phone easy 615 Big Button SIM Free Mobile ...   10.0             44
Samsung i9195 Galaxy S4 mini Mobiltelefon            10.0             40
Телефон LG E400 Optimus L3 Black              10.0             39
Телефон LG P970 Optimus Black Titanium B...   10.0             38

# 4. Build a collaborative filtering model using SVD.

In [79]:
# arranging columns in the order of user id,item id and rating to be fed in the svd
columns_titles = ['author','product','score']
vs_rev = df2.reindex(columns=columns_titles)

In [80]:
# Keep only 5000 data samples. Use random state=612
vs_data = vs_rev.sample(n=5000, random_state=612)

In [81]:
# 4. Build a collaborative filtering model using SVD. 
reader = Reader(rating_scale=(1, 10))
data = Dataset.load_from_df(vs_data,reader = reader)

In [82]:
trainset = data.build_full_trainset()

In [83]:
trainset.ur


defaultdict(list,
            {0: [(0, 8.0)],
             1: [(1, 8.0)],
             2: [(2, 10.0)],
             3: [(3, 8.0)],
             4: [(4, 8.0)],
             5: [(5, 10.0)],
             6: [(6, 10.0)],
             7: [(7, 10.0)],
             8: [(8, 10.0)],
             9: [(9, 10.0)],
             10: [(10, 10.0)],
             11: [(11, 8.0)],
             12: [(12, 8.0)],
             13: [(13, 8.0)],
             14: [(14, 2.0)],
             15: [(15, 10.0)],
             16: [(16, 6.0)],
             17: [(17, 6.0)],
             18: [(18, 8.0)],
             19: [(19, 10.0)],
             20: [(20, 8.0)],
             21: [(21, 10.0)],
             22: [(22, 8.0)],
             23: [(23, 8.0)],
             24: [(24, 10.0)],
             25: [(25, 10.0)],
             26: [(26, 10.0)],
             27: [(27, 2.0)],
             28: [(28, 8.0)],
             29: [(29, 6.0), (2241, 4.0)],
             30: [(30, 8.0)],
             31: [(31, 10.0), (218, 10.0), (21

In [84]:
algo = SVD()
algo.fit(trainset)

In [85]:
# Than predict ratings for all pairs (u, i) that are NOT in the training set.
testset = trainset.build_anti_testset()

In [86]:
predictions = algo.test(testset)

In [87]:
predictions

[Prediction(uid='Frota', iid='SONY ERICSSON Vivaz', r_ui=7.7382, est=7.6496287978302115, details={'was_impossible': False}),
 Prediction(uid='Frota', iid='HP Slate 6 Voice Tab (6inch,16GB,Wi-Fi+3G+Voice Calling), Graphite', r_ui=7.7382, est=7.359374529801991, details={'was_impossible': False}),
 Prediction(uid='Frota', iid='Samsung Wave 3 GT-S8600', r_ui=7.7382, est=7.829761401088123, details={'was_impossible': False}),
 Prediction(uid='Frota', iid='Apple iPhone 3GS 8Go Noir', r_ui=7.7382, est=7.658618720713803, details={'was_impossible': False}),
 Prediction(uid='Frota', iid='Samsung Galaxy S II i9100 DualCore Smartphone (10.9 cm (4.3 Zoll) Super-Amoled Plus Display, Android 4.0 oder höher, 8 MP Full-HD Kamera, 2 MP Frontkamera) noble-black', r_ui=7.7382, est=7.8156548463118725, details={'was_impossible': False}),
 Prediction(uid='Frota', iid='Samsung Galaxy S II i9100 16 GB Unlocked GSM Smartphone, Noble Black', r_ui=7.7382, est=7.80024795475928, details={'was_impossible': False}),


Build a collaborative filtering model using kNNWithMeans from surprise using User based model


In [89]:
reader = Reader(rating_scale=(1, 10))
data_U = Dataset.load_from_df(vs_data,reader = reader)

In [90]:
trainset_U, testset_U = train_test_split(data_U, test_size=.15)


In [91]:
# Use user_based true/false to switch between user-based or item-based collaborative filtering
algo = KNNWithMeans(k=50, sim_options={'name': 'pearson_baseline', 'user_based': True})
algo.fit(trainset_U)

Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.


In [92]:
# we can now query for specific predicions
uid = 'Frances DeSimone'  # raw user id
iid = 'Samsung Galaxy Star Pro DUOS S7262 Unlocked Ce.'  # raw item id

In [93]:
# get a prediction for specific users and items.
pred = algo.predict(uid, iid, verbose=True)

user: Frances DeSimone item: Samsung Galaxy Star Pro DUOS S7262 Unlocked Ce. r_ui = None   est = 7.74   {'was_impossible': True, 'reason': 'User and/or item is unknown.'}


5. Evaluate the collaborative model. Print RMSE value.

In [94]:
print("SVD Model : Test Set")
accuracy.rmse(predictions, verbose=True)

SVD Model : Test Set
RMSE: 0.3719


0.37190047749442706

6. Predict score (average rating) for test users. 

In [95]:
# run the trained model against the testset
test_pred_U = algo.test(testset_U)

In [96]:
test_pred_U

[Prediction(uid='FRANCIS_BRAGA2008', iid='Motorola V9', r_ui=10.0, est=7.736941176470588, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid='samir1917', iid='Huawei IDEOS X3', r_ui=10.0, est=7.736941176470588, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid='RobTouchPro2', iid='HTC Touch Pro2 CDMA', r_ui=9.0, est=7.736941176470588, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid='dennyboomboom', iid='Sony Phones Xperia Go Sim Free Mobile Phone - Black', r_ui=2.0, est=7.736941176470588, details={'was_impossible': True, 'reason': 'User and/or item is unknown.'}),
 Prediction(uid='emanetcan', iid='Nokia 3720 Classic cep telefonu', r_ui=8.0, est=8.0, details={'actual_k': 0, 'was_impossible': False}),
 Prediction(uid='Ð\x9cакÑ\x81именко ?Ñ\x80ина Ð\x9fеÑ\x82Ñ\x80?вна', iid='Nokia 5230 XpressMusic', r_ui=10.0, est=7.736941176470588, details={'was

7. Report your findings and inferences. 

Overall data is highly skewed towards 'Amazon customers' 
Most of the authors have given the rating of '10' or '8'

8. Try and recommend top 5 products for test users. 

In [97]:
def get_top_n(predictions, n=5):
    # First map the predictions to each user.
    top_n = defaultdict(list)
    for uid, iid, true_r, est, _ in predictions:
        top_n[uid].append((iid, est))

    # Then sort the predictions for each user and retrieve the k highest ones.
    for uid, user_ratings in top_n.items():
        user_ratings.sort(key=lambda x: x[1], reverse=True)
        top_n[uid] = user_ratings[:n]

    return top_n

In [98]:
# 8. Try and recommend top 5 products for test users
top_n = get_top_n(predictions, n=5)

In [99]:
top_n

defaultdict(list,
            {'Frota': [('HTC Desire S', 8.900019180452748),
              ('Nokia 5800', 8.760514383891639),
              ('LG Optimus L7 P705 Black', 8.72791664151785),
              ('Nokia E63', 8.7237379653512),
              ('Apple iPhone 3GS 32Gb', 8.70911758984763)],
             'papounet59': [('HTC Desire S', 8.910146476850782),
              ('LG Optimus One P500', 8.76772604504313),
              ('Samsung GT-S5230 Star', 8.762989153287107),
              ('Samsung Galaxy S II White', 8.755018452974916),
              ('LG Optimus L7 P705 Black', 8.705643414245184)],
             'JAGAJIVAN': [('Samsung Galaxy S II White', 9.191390703757865),
              ('Samsung Jet S8000', 9.039213527975026),
              ('HTC Desire S', 9.036751881319423),
              ('Nokia 5800', 8.892026355121496),
              ('Nokia E63', 8.877524996602144)],
             'uncommon13': [('Apple iPhone 4 16Gb', 8.972416102084212),
              ('HTC Desire S', 8.86196709

9. Try other techniques (Example: cross validation) to get better results. 

In [100]:
cross_validate(algo, data, measures=['RMSE'], cv=3, verbose=False)

Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.
Estimating biases using als...
Computing the pearson_baseline similarity matrix...
Done computing similarity matrix.


{'test_rmse': array([2.54097559, 2.50849413, 2.6213206 ]),
 'fit_time': (0.3203132152557373, 0.21360397338867188, 0.17496514320373535),
 'test_time': (0.006541013717651367,
  0.005762815475463867,
  0.005630970001220703)}

Observation :
    RMSE of SVD model is lower than for cross validation.

10. In what business scenario you should use popularity based Recommendation Systems ?

It does not suffer from cold start problems which means on day 1 of the business also it can recommend products on various different filters.

There is no need for the user's historical data.

Popularity based recommendation system relies on the popularity,trends and frequency counts of which items were most purchased.

It is used buy the travel companies selling holiday packages in a season, by Google News and other news websites to show Top Stories with images.

11. In what business scenario you should use CF based Recommendation Systems ?

.Collaborative Filtering is used to building intelligent recommender systems that can learn to give better recommendations as more information about users is collected. It is a  personalised recommender system , recommendations are made based on the past behaviour of the user. Most websites like Amazon, YouTube, and Netflix use collaborative filtering as a part of their sophisticated recommendation system.

12. What other possible methods can you think of which can further improve the recommendation for different users ?

Apart from Popularity and Collaborative Filtering , Content-based, Demographic, Utility based, Knowledge based and Hybrid recommendation system can be used as per the user needs.